platform/api/esx/archive.go

```diff
@@ -0,0 +1,99 @@
+// Copyright 2017 CoreOS, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package esx
+
+import (
+	"archive/tar"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path"
+
+	"github.com/vmware/govmomi/ovf"
+)
+
+type Archive struct {
+	path string
+}
+
+type ArchiveEntry struct {
+	io.Reader
+	f *os.File
+}
+
+func (t *ArchiveEntry) Close() error {
+	return t.f.Close()
+}
+
+func (t *Archive) ReadOvf(fpath string) ([]byte, error) {
+	r, _, err := t.Open(fpath)
+	if err != nil {
+		return nil, err
+	}
+	defer r.Close()
+
+	return ioutil.ReadAll(r)
+}
+
+func (t *Archive) ReadEnvelope(fpath string) (*ovf.Envelope, error) {
+	if fpath == "" {
+		return nil, nil
+	}
+
+	r, _, err := t.Open(fpath)
+	if err != nil {
+		return nil, err
+	}
+	defer r.Close()
+
+	e, err := ovf.Unmarshal(r)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse ovf: %s", err.Error())
+	}
+
+	return e, nil
+}
+
+func (t *Archive) Open(name string) (io.ReadCloser, int64, error) {
+	f, err := os.Open(t.path)
+	defer f.Close()
+	if err != nil {
+		return nil, 0, err
+	}
+
+	r := tar.NewReader(f)
+
+	for {
+		h, err := r.Next()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return nil, 0, err
+		}
+
+		matched, err := path.Match(name, path.Base(h.Name))
+		if err != nil {
+			return nil, 0, err
+		}
+
+		if matched {
+			return &ArchiveEntry{r, f}, h.Size, nil
```

Comment:
You're closing `f` on return from this function, so I'm surprised this works.  It should only be closed on error.